controller/backup_controller.go

```diff
@@ -587,6 +591,47 @@ func (bc *BackupController) hasMonitor(backupName string) *engineapi.BackupMonit
 	return bc.monitors[backupName]
 }
 
+//getBackupVolumeRecurringJobs get recurring jobs in the volume labels and recurring jobs of groups in the volume labels.
+func (bc *BackupController) getBackupVolumeRecurringJobs(volume *longhorn.Volume) string {
+	volumeRecurringJobSpecList := make(map[string]longhorn.RecordVolumeRecurringJob)
+
+	allRecurringJobs, err := bc.ds.ListRecurringJobsRO()
+	if err != nil {
+		bc.logger.WithError(err).Warn("Cannot list all recurring jobs")
+	}
+	// distinguish volume labels between job and group.
+	jobs := datastore.MarshalLabelToVolumeRecurringJob(volume.Labels)
+	for recurringJobName, recurringJob := range allRecurringJobs {
+		for jobName, job := range jobs {
+			if job.IsGroup {
+				if !util.Contains(recurringJob.Spec.Groups, jobName) {
+					continue
+				}
+				if recordJob, exist := volumeRecurringJobSpecList[recurringJobName]; !exist {
+					volumeRecurringJobSpecList[recurringJobName] = longhorn.RecordVolumeRecurringJob{JobSpec: recurringJob.Spec, FromGroup: []string{jobName}}
+				} else if !util.Contains(volumeRecurringJobSpecList[recurringJobName].FromGroup, jobName) {
+					recordJob.FromGroup = append(recordJob.FromGroup, jobName)
+					volumeRecurringJobSpecList[recurringJobName] = recordJob
+				}
+			} else if recurringJobName == jobName {
+				// save single job information if it is not a group.
+				if recordJob, exist := volumeRecurringJobSpecList[recurringJobName]; !exist {
+					volumeRecurringJobSpecList[recurringJobName] = longhorn.RecordVolumeRecurringJob{JobSpec: recurringJob.Spec, FromJob: true, FromGroup: []string{}}
+				} else {
+					recordJob.FromJob = true
+					volumeRecurringJobSpecList[recurringJobName] = recordJob
+				}
+			}
```

Comment:
Add more comments or try to abstract some functions for this nested part. It's not easy-to-understand.

Reply:
Added, and separate adding recurring jobs from group into another function.